File: DevExpress VCL/ExpressQuantumTreeList 4/Demos/Delphi/SummariesDemo/SummariesDemoMain.pas
unit SummariesDemoMain;

{$I ..\cxVer.inc}

interface

uses
  Windows, Messages, SysUtils, {$IFDEF DELPHI6}Variants, {$ENDIF}Classes, Graphics, Controls, Forms,
  Dialogs, cxControls, cxLookAndFeels, ActnList, ImgList, Menus, ComCtrls,
  StdCtrls, DemoBasicMain, cxContainer, cxEdit, cxTextEdit, cxStyles, cxTL,
  cxInplaceContainer, cxTLData, cxDBTL, cxMaskEdit, cxCalendar,
  cxCurrencyEdit, cxDBLookupComboBox, cxDBEditRepository,
  cxEditRepositoryItems, cxImage, cxMemo,
  cxLookAndFeelPainters, cxButtons, cxDropDownEdit, cxTimeEdit,
  cxImageComboBox, cxCalc, cxSpinEdit, cxCustomData, Grids, DBGrids,
  cxGraphics;

type
  TSummariesDemoMainForm = class(TDemoBasicMainForm)
    miGridLookFeel: TMenuItem;
    miKind: TMenuItem;
    miFlat: TMenuItem;
    miStandard: TMenuItem;
    miUltraFlat: TMenuItem;
    miNativeStyle: TMenuItem;
    miSeparator2: TMenuItem;
    tlOrders: TcxDBTreeList;
    tlOrdersPurchaseDate: TcxDBTreeListColumn;
    tlOrdersTime: TcxDBTreeListColumn;
    tlOrdersPaymentType: TcxDBTreeListColumn;
    tlOrdersPaymentAmount: TcxDBTreeListColumn;
    tlOrdersQuantity: TcxDBTreeListColumn;
    tlOrdersCustomer: TcxDBTreeListColumn;
    tlOrdersCar: TcxDBTreeListColumn;
    cxEditRepository: TcxEditRepository;
    eriCarLookup: TcxEditRepositoryLookupComboBoxItem;
    eriCarInfoPopup: TcxEditRepositoryPopupItem;
    N1: TMenuItem;
    miSummaries: TMenuItem;
    miIgnoreNullValues: TMenuItem;
    miOnlyVisaPaymentType: TMenuItem;
    miCustomizeSummaries: TMenuItem;
    miSelectedRecordsOnly: TMenuItem;
    N2: TMenuItem;
    N3: TMenuItem;
    miMultiselect: TMenuItem;
    Office111: TMenuItem;
    procedure LookAndFeelChange(Sender: TObject);
    procedure FormShow(Sender: TObject);
    procedure miCustomizeSummariesClick(Sender: TObject);
    procedure miIgnoreNullValuesClick(Sender: TObject);
    procedure miMultiselectClick(Sender: TObject);
    procedure miOnlyVisaPaymentTypeClick(Sender: TObject);
    procedure miSelectedRecordsOnlyClick(Sender: TObject);
    procedure tlOrdersAfterSummary(ASender: TcxDataSummary);
    procedure tlOrdersSummary(ASender: TcxDataSummaryItems;
      Arguments: TcxSummaryEventArguments;
      var OutArguments: TcxSummaryEventOutArguments);
    procedure tlOrdersSelectionChanged(Sender: TObject);
  private
    FCalcForVisaOnly: Boolean;
    FCalcSelectedOnly: Boolean;
  end;

var
  SummariesDemoMainForm: TSummariesDemoMainForm;

implementation

uses SummariesDemoData, ShellAPI, SummariesDemoEditSummary, DB, cxDBData;

{$R *.dfm}

procedure TSummariesDemoMainForm.LookAndFeelChange(Sender: TObject);
begin
  if TMenuItem(Sender).Tag > 3 then
  begin
    cxLookAndFeelController.NativeStyle :=
      not cxLookAndFeelController.NativeStyle;
    TMenuItem(Sender).Checked := cxLookAndFeelController.NativeStyle;
  end
  else
  begin
    TMenuItem(Sender).Checked := True;
    cxLookAndFeelController.Kind := TcxLookAndFeelKind(TMenuItem(Sender).Tag);
    cxLookAndFeelController.NativeStyle := False;
    miNativeStyle.Checked := False;
  end;
end;

procedure TSummariesDemoMainForm.FormShow(Sender: TObject);
begin
  FCalcForVisaOnly := False;
  FCalcSelectedOnly := False;
  miCustomizeSummariesClick(nil);
{ remove/add the closing brace on this line to disable/enable the following code

  ShowMessage('WARNING: tutorial not completed. First, please apply the steps '+
              'shown in the doc file');

//}
end;

procedure TSummariesDemoMainForm.miCustomizeSummariesClick(
  Sender: TObject);
begin
  if not Assigned(SummariesDemoEditSummaryForm) then
  begin
    SummariesDemoEditSummaryForm := TSummariesDemoEditSummaryForm.Create(Application);
    SummariesDemoEditSummaryForm.TreeList := tlOrders;
  end;
  SummariesDemoEditSummaryForm.Show;
end;

procedure TSummariesDemoMainForm.miIgnoreNullValuesClick(Sender: TObject);
begin
  TMenuItem(Sender).Checked := not TMenuItem(Sender).Checked;
  tlOrders.OptionsData.SummaryNullIgnore := TMenuItem(Sender).Checked;
end;

procedure TSummariesDemoMainForm.miMultiselectClick(Sender: TObject);
begin
  TMenuItem(Sender).Checked := not TMenuItem(Sender).Checked;
  tlOrders.OptionsSelection.MultiSelect := TMenuItem(Sender).Checked;
end;

procedure TSummariesDemoMainForm.miOnlyVisaPaymentTypeClick(
  Sender: TObject);
begin
  TMenuItem(Sender).Checked := not TMenuItem(Sender).Checked;
  tlOrders.DataController.Summary.BeginUpdate;
  try
    FCalcForVisaOnly := TMenuItem(Sender).Checked;
  finally
    tlOrders.DataController.Summary.EndUpdate;
  end;
end;

procedure TSummariesDemoMainForm.miSelectedRecordsOnlyClick(
  Sender: TObject);
begin
  TMenuItem(Sender).Checked := not TMenuItem(Sender).Checked;
  tlOrders.DataController.Summary.BeginUpdate;
  miOnlyVisaPaymentType.Enabled := not TMenuItem(Sender).Checked;
  miIgnoreNullValues.Enabled := not TMenuItem(Sender).Checked;
  FCalcForVisaOnly := miOnlyVisaPaymentType.Enabled and
    miOnlyVisaPaymentType.Checked;
  try
    FCalcSelectedOnly := TMenuItem(Sender).Checked;
  finally
    tlOrders.DataController.Summary.EndUpdate;
  end;
end;

procedure TSummariesDemoMainForm.tlOrdersAfterSummary(
  ASender: TcxDataSummary);

  function SummaryKindToStr(AKind: TcxSummaryKind): string;
  begin
    case AKind of
      skSum:
        Result := 'SUM';
      skMin:
        Result := 'MIN';
      skMax:
        Result := 'MAX';
      skCount:
        Result := 'Count';
      skAverage:
        Result := 'AVG';
    else
      Result := ''
    end;
  end;

  function GetSQLCondition: string;
  var
    I: Integer;
  begin
    Result := ' (';
    for I := 0 to tlOrders.SelectionCount - 1 do
    begin
      Result := Result +
        VarToStr(TcxTreeListDataNode(tlOrders.Selections[I]).KeyValue) + ', ';
    end;
    Delete(Result, Length(Result) - 1, 2);
    Result := Result + ')';
  end;
               
Const
  ACalculatedTableName = 'Orders';
  ABackSeparator = '''';
  AFrontSeparator = '''';

var
  I: Integer;
  AKeyFieldName, AFieldName, ASQL: string;
  ASummaryItems: TcxDataFooterSummaryItems;
  ASummaryItem: TcxDBDataSummaryItem;
begin
{ remove/add the closing brace on this line to disable/enable the following code}

  if not FCalcSelectedOnly then Exit;
  AKeyFieldName := tlOrders.DataController.KeyField;
  ASummaryItems := ASender.FooterSummaryItems;
  ASQL := 'Select ';
  for I := 0 to ASummaryItems.Count - 1 do
  begin
    ASummaryItem := TcxDBDataSummaryItem(ASummaryItems[I]);
    AFieldName := ASummaryItem.FieldName;
    if AFieldName = '' then
      AFieldName :=
        (ASummaryItem.ItemLink as TcxDBTreeListColumn).DataBinding.FieldName;
    AFieldName := ABackSeparator + ACalculatedTableName + AFrontSeparator +
       '.'+ ABackSeparator + AFieldName + AFrontSeparator;
    ASQL := ASQL + SummaryKindToStr(ASummaryItem.Kind) + '(' + AFieldName + '), ';
  end;
  Delete(ASQL, Length(ASQL) - 1, 2);
  ASQL := ASQL + ' From ' + ACalculatedTableName;
  with SummariesDemoDataDM do
  begin
    qryOrdersSummary.Close;
    qryOrdersSummary.SQL.Clear;
    qryOrdersSummary.SQL.Add(ASQL);
    if tlOrders.SelectionCount > 0 then
    begin
      qryOrdersSummary.SQL.Add('where ' + AKeyFieldName + ' in ');
      qryOrdersSummary.SQL.Add(GetSQLCondition);
    end;
    qryOrdersSummary.Open;
    for I := 0 to ASummaryItems.Count - 1 do
      ASender.FooterSummaryValues[I] := qryOrdersSummary.Fields[I].Value;
    qryOrdersSummary.Close;
  end;
  tlOrders.Invalidate;
//}
end;

procedure TSummariesDemoMainForm.tlOrdersSummary(
  ASender: TcxDataSummaryItems; Arguments: TcxSummaryEventArguments;
  var OutArguments: TcxSummaryEventOutArguments);
begin
{ remove/add the closing brace on this line to disable/enable the following code}

  if not FCalcForVisaOnly then Exit;
  if ASender.DataController.Values[Arguments.RecordIndex, tlOrdersPaymentType.ItemIndex] <> 'Visa' then
    OutArguments.Done := True;

//}
end;

procedure TSummariesDemoMainForm.tlOrdersSelectionChanged(Sender: TObject);
begin
{ remove/add the closing brace on this line to disable/enable the following code}

  tlOrders.DataController.Summary.Calculate;

//}
end;

end.

File: DevExpress VCL/ExpressQuantumTreeList 4/Demos/CBuilder/SummariesDemo/SummariesDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#include "..\cxDemosBCB.inc"
#pragma hdrstop

#include "SummariesDemoMain.h"
#include "SummariesDemoData.h"
#include "SummariesDemoEditSummary.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxClasses"
#pragma link "cxControls"
#pragma link "cxCustomData"
#pragma link "cxData"
#pragma link "cxDBData"
#pragma link "cxEdit"
#pragma link "cxFilter"
#pragma link "cxGraphics"
#pragma link "cxStyles"
#pragma link "cxLookAndFeels"
#pragma link "DemoBasicMain"
#pragma link "cxCalc"
#pragma link "cxDBEditRepository"
#pragma link "cxDBLookupComboBox"
#pragma link "cxDBTL"
#pragma link "cxEditRepositoryItems"
#pragma link "cxImageComboBox"
#pragma link "cxInplaceContainer"
#pragma link "cxMaskEdit"
#pragma link "cxSpinEdit"
#pragma link "cxTimeEdit"
#pragma link "cxTL"
#pragma link "cxTLData"
#pragma link "cxCalendar"
#pragma resource "*.dfm"
TSummariesDemoMainForm *SummariesDemoMainForm;
//---------------------------------------------------------------------------
__fastcall TSummariesDemoMainForm::TSummariesDemoMainForm(TComponent* Owner)
  : TDemoBasicMainForm(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TSummariesDemoMainForm::FormShow(TObject *Sender)
{
  FCalcForVisaOnly = false;
  FCalcSelectedOnly = false;
  miCustomizeSummariesClick(NULL);
/* remove/add the closing slash on this line to disable/enable the following code *

  ShowMessage("WARNING: tutorial not completed. First, please apply the steps "
    "shown in the doc file");

//*/
}
//---------------------------------------------------------------------------

void __fastcall TSummariesDemoMainForm::LookAndFeelChange(TObject *Sender)
{
  if (((TMenuItem*)Sender)->Tag > 3){
    cxLookAndFeelController->NativeStyle =
      !cxLookAndFeelController->NativeStyle;
    ((TMenuItem*)Sender)->Checked = cxLookAndFeelController->NativeStyle;
  }
  else{
    ((TMenuItem*)Sender)->Checked = true;
    cxLookAndFeelController->Kind = (TcxLookAndFeelKind)((TMenuItem*)Sender)->Tag;
    cxLookAndFeelController->NativeStyle = false;
    miNativeStyle->Checked = false;
  }
}
//---------------------------------------------------------------------------

void __fastcall TSummariesDemoMainForm::miCustomizeSummariesClick(TObject *Sender)
{
  if (SummariesDemoEditSummaryForm == NULL){
    SummariesDemoEditSummaryForm = new TSummariesDemoEditSummaryForm(Application);
    SummariesDemoEditSummaryForm->TreeList = tlOrders;
  }
  SummariesDemoEditSummaryForm->Show();
}
//---------------------------------------------------------------------------

void __fastcall TSummariesDemoMainForm::miIgnoreNullValuesClick(TObject *Sender)
{
  ((TMenuItem*)Sender)->Checked = !((TMenuItem*)Sender)->Checked;
  tlOrders->OptionsData->SummaryNullIgnore = ((TMenuItem*)Sender)->Checked;
}
//---------------------------------------------------------------------------

void __fastcall TSummariesDemoMainForm::miMultiselectClick(TObject *Sender)
{
  ((TMenuItem*)Sender)->Checked = !((TMenuItem*)Sender)->Checked;
  tlOrders->OptionsSelection->MultiSelect = ((TMenuItem*)Sender)->Checked;
}
//---------------------------------------------------------------------------

void __fastcall TSummariesDemoMainForm::miOnlyVisaPaymentTypeClick(TObject *Sender)
{
  ((TMenuItem*)Sender)->Checked = !((TMenuItem*)Sender)->Checked;
  tlOrders->DataController->Summary->BeginUpdate();
  __try{
    FCalcForVisaOnly = ((TMenuItem*)Sender)->Checked;
  }
  __finally{
    tlOrders->DataController->Summary->EndUpdate();
  }
}
//---------------------------------------------------------------------------

void __fastcall TSummariesDemoMainForm::miSelectedRecordsOnlyClick(TObject *Sender)
{
  ((TMenuItem*)Sender)->Checked = !((TMenuItem*)Sender)->Checked;
  tlOrders->DataController->Summary->BeginUpdate();
  __try{
    miOnlyVisaPaymentType->Enabled = !((TMenuItem*)Sender)->Checked;
    miIgnoreNullValues->Enabled = !((TMenuItem*)Sender)->Checked;
    FCalcForVisaOnly = miOnlyVisaPaymentType->Enabled &&
    miOnlyVisaPaymentType->Checked;
    FCalcSelectedOnly = ((TMenuItem*)Sender)->Checked;
  }
  __finally{
    tlOrders->DataController->Summary->EndUpdate();
  }
}
//---------------------------------------------------------------------------

void __fastcall TSummariesDemoMainForm::tlOrdersAfterSummary(
      TcxDataSummary *ASender)
{
const String ACalculatedTableName = "Orders";
#if BCB6
  const String ABackSeparator = "'";
  const String AFrontSeparator =  "'";
#else
  const String ABackSeparator = "'";
  const String AFrontSeparator = "'";
#endif

/* remove/add the closing slash on this line to disable/enable the following code */

  if (!FCalcSelectedOnly) return;
  String AKeyFieldName = tlOrders->DataController->KeyField;
  TcxDataFooterSummaryItems *ASummaryItems = ASender->FooterSummaryItems;
  String ASQL = "Select ";
  TcxDBDataSummaryItem *ASummaryItem;
  String AFieldName;
  for (int I = 0; I < ASummaryItems->Count; I++){
    ASummaryItem = (TcxDBDataSummaryItem*)ASummaryItems->Items[I];
    AFieldName = ASummaryItem->FieldName;
    if (AFieldName == "")
      AFieldName =
        ((TcxDBTreeListColumn*)ASummaryItem->ItemLink)->DataBinding->FieldName;
    AFieldName = ABackSeparator + ACalculatedTableName + AFrontSeparator +
       "."+ ABackSeparator + AFieldName + AFrontSeparator;
    ASQL += SummaryKindToStr(ASummaryItem->Kind) + "(" + AFieldName + "), ";
  }
  ASQL.Delete(ASQL.Length() - 1, 2);
  ASQL += " From " + ACalculatedTableName;
  SummariesDemoDataDM->qryOrdersSummary->Close();
  SummariesDemoDataDM->qryOrdersSummary->SQL->Clear();
  SummariesDemoDataDM->qryOrdersSummary->SQL->Add(ASQL);
  if (tlOrders->SelectionCount > 0){
    SummariesDemoDataDM->qryOrdersSummary->SQL->Add("where " + AKeyFieldName + " in ");
    SummariesDemoDataDM->qryOrdersSummary->SQL->Add(GetSQLCondition());
  }
  SummariesDemoDataDM->qryOrdersSummary->Open();
  for (int I = 0; I < ASummaryItems->Count; I++)
    ASender->FooterSummaryValues[I] =
      SummariesDemoDataDM->qryOrdersSummary->Fields->Fields[I]->Value;
  SummariesDemoDataDM->qryOrdersSummary->Close();

//*/
}

String TSummariesDemoMainForm::SummaryKindToStr(TcxSummaryKind AKind)
{
  String Result;
  switch (AKind){
    case skSum:
      Result = "SUM"; break;
    case skMin:
      Result = "MIN"; break;
    case skMax:
      Result = "MAX"; break;
    case skCount:
      Result = "Count"; break;
    case skAverage:
      Result = "AVG"; break;
    default:
      Result = "";
  }
  return Result;
}

String TSummariesDemoMainForm::GetSQLCondition()
{
    String Result = " (";
    for (int I = 0; I < tlOrders->SelectionCount; I++){
      Result += VarToStr(((TcxTreeListDataNode*)tlOrders->Selections[I])->KeyValue) + ", ";
    }
    Result.Delete(Result.Length() - 1, 2);
    Result += ")";
    return Result;
}
//---------------------------------------------------------------------------

void __fastcall TSummariesDemoMainForm::tlOrdersSummary(
      TcxDataSummaryItems *ASender, TcxSummaryEventArguments &Arguments,
      TcxSummaryEventOutArguments &OutArguments)
{
/* remove/add the closing slash on this line to disable/enable the following code */

  if (!FCalcForVisaOnly) return;
  if (VarToStr(ASender->DataController->Values[Arguments.RecordIndex][tlOrdersPaymentType->ItemIndex]) != "Visa")
    OutArguments.Done = true;

//*/
}
//---------------------------------------------------------------------------

void __fastcall TSummariesDemoMainForm::tlOrdersSelectionChanged(TObject *Sender)
{
/* remove/add the closing slash on this line to disable/enable the following code */

  tlOrders->DataController->Summary->Calculate();

//*/
}
//---------------------------------------------------------------------------


